xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java

```diff
@@ -0,0 +1,224 @@
+    DownstreamTlsContext downstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildDownstreamTlsContextFromFilenames(
+            BAD_SERVER_KEY_FILE,
+            BAD_SERVER_PEM_FILE,
+            SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+    createListenerUpdate(server.getPort(), downstreamTlsContext, listenerWatcher);
+    try {
+      buildClientAndTest(upstreamTlsContext, "foo.test.google.fr", "buddy", server.getPort());
+      fail("exception expected");
+    } catch (StatusRuntimeException sre) {
+      assertThat(sre).hasCauseThat().isInstanceOf(SSLHandshakeException.class);
+      assertThat(sre).hasCauseThat().hasMessageThat().isEqualTo("General OpenSslEngine problem");
+    }
+  }
+
+  private XdsClient.ListenerWatcher mtlsCommonTest(UpstreamTlsContext upstreamTlsContext)
+      throws IOException {
+    DownstreamTlsContext downstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildDownstreamTlsContextFromFilenames(
+            SecretVolumeSslContextProviderTest.SERVER_1_KEY_FILE,
+            SecretVolumeSslContextProviderTest.SERVER_1_PEM_FILE,
+            SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+
+    XdsClient.ListenerWatcher listenerWatcher = getXdsServer(downstreamTlsContext);
+    buildClientAndTest(upstreamTlsContext, "foo.test.google.fr", "buddy", server.getPort());
+    return listenerWatcher;
+  }
+
+  private XdsClient.ListenerWatcher getXdsServer(DownstreamTlsContext downstreamTlsContext)
+      throws IOException {
+    int port = findFreePort();
+    XdsServerBuilder builder =
+        XdsServerBuilder.forPort(port).addService(new SimpleServiceImpl());
+    final XdsClientWrapperForServerSds xdsClientWrapperForServerSds =
+        createXdsClientWrapperForServerSds(port, downstreamTlsContext);
+    SdsProtocolNegotiators.ServerSdsProtocolNegotiator serverSdsProtocolNegotiator =
+        new SdsProtocolNegotiators.ServerSdsProtocolNegotiator(xdsClientWrapperForServerSds);
+    server = cleanupRule.register(builder.buildServer(serverSdsProtocolNegotiator)).start();
+    return xdsClientWrapperForServerSds.getListenerWatcher();
+  }
+
+  /** Creates XdsClientWrapperForServerSds: also used by other classes. */
+  public static XdsClientWrapperForServerSds createXdsClientWrapperForServerSds(
+      int port, DownstreamTlsContext downstreamTlsContext) {
+    XdsClient mockXdsClient = mock(XdsClient.class);
+    XdsClientWrapperForServerSds xdsClientWrapperForServerSds =
+        new XdsClientWrapperForServerSds(port, mockXdsClient, null);
+    createListenerUpdate(port, downstreamTlsContext,
+        xdsClientWrapperForServerSds.getListenerWatcher());
+    return xdsClientWrapperForServerSds;
+  }
+
+  private static int findFreePort() throws IOException {
+    try (ServerSocket socket = new ServerSocket(0)) {
+      socket.setReuseAddress(true);
+      return socket.getLocalPort();
+    }
+  }
+
+  private static void createListenerUpdate(
```

Comment:
the method name doesn't match to what it actually does. it makes the reader very confusing.